### ISSUE TITLE
Dataloader not working when num_workers > 0

### ISSUE THREAD

**Issue by evorta** [apache/incubator-mxnet#13792]
When I train a ssd neural network using gluoncv 1.3.0 or 1.4.0 and mxnet. I use mxnet's dataloader:

```
from mxnet import gluon

def get_dataloader(net, train_dataset, data_shape, batch_size, num_workers):
    from gluoncv.data.batchify import Tuple, Stack, Pad
    from gluoncv.data.transforms.presets.ssd import SSDDefaultTrainTransform
    width, height = data_shape, data_shape
    # use fake data to generate fixed anchors for target generation
    with autograd.train_mode():
        _, _, anchors = net(mx.nd.zeros((1, 3, height, width)))
    batchify_fn = Tuple(Stack(), Stack(), Stack())  # stack image, cls_targets, box_targets
    train_loader = gluon.data.DataLoader(
        train_dataset.transform(SSDDefaultTrainTransform(width, height, anchors)),
        batch_size, True, batchify_fn=batchify_fn, last_batch='rollover', num_workers=num_workers)
    return train_loader

train_data = get_dataloader(net, train_dataset, 512, BATCH_SIZE, NUM_WORKERS)
```

To turn my dataset into batches ready for processing in a gluoncv training loop. I can train when NUM_WORKERS = 0, it works fine, not using 100% of the GPU but at least some, as I saturate a single core and only 10% of my GPU.

If I increase the num_workers > 0 the system uses 100% of CPU and refuses to continue, when ctrl-c'ding the program it shows it is stuck (must kill -9 to stop it) in:

```
> Traceback (most recent call last):
>   File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
>     self.run()
>   File "/usr/lib/python2.7/multiprocessing/process.py", line 114, in run
>     self._target(*self._args, **self._kwargs)
>   File "/usr/lib/python2.7/multiprocessing/pool.py", line 113, in worker
>     result = (True, func(*args, **kwds))
>   File "/usr/local/lib/python2.7/dist-packages/mxnet-1.5.0-py2.7.egg/mxnet/gluon/data/dataloader.py", line 394, in _worker_fn
>     batch = batchify_fn([_worker_dataset[i] for i in samples])
>   File "/usr/local/lib/python2.7/dist-packages/mxnet-1.5.0-py2.7.egg/mxnet/gluon/data/dataset.py", line 124, in __getitem__
>     return self._fn(*item)
>   File "/usr/local/lib/python2.7/dist-packages/gluoncv/data/transforms/presets/ssd.py", line 176, in __call__
>     gt_bboxes = mx.nd.array(bbox[np.newaxis, :, :4])
>   File "/usr/local/lib/python2.7/dist-packages/mxnet-1.5.0-py2.7.egg/mxnet/ndarray/utils.py", line 146, in array
>     return _array(source_array, ctx=ctx, dtype=dtype)
>   File "/usr/local/lib/python2.7/dist-packages/mxnet-1.5.0-py2.7.egg/mxnet/ndarray/ndarray.py", line 2496, in array
>     arr[:] = source_array
>   File "/usr/local/lib/python2.7/dist-packages/mxnet-1.5.0-py2.7.egg/mxnet/ndarray/ndarray.py", line 451, in __setitem__
>     self._set_nd_basic_indexing(key, value)
>   File "/usr/local/lib/python2.7/dist-packages/mxnet-1.5.0-py2.7.egg/mxnet/ndarray/ndarray.py", line 717, in _set_nd_basic_indexing
>     self._sync_copyfrom(value)
>   File "/usr/local/lib/python2.7/dist-packages/mxnet-1.5.0-py2.7.egg/mxnet/ndarray/ndarray.py", line 883, in _sync_copyfrom
>     ctypes.c_size_t(source_array.size)))
> KeyboardInterrupt
> python: malloc.c:3722: _int_malloc: Assertion `(unsigned long) (size) >= (unsigned long) (nb)' failed.
> 
> 
> ^C^CKilled
```

## Environment info (Required)

pip list | grep gluoncv
gluoncv                       0.4.0b20190107
pip list | grep mxnet
mxnet                         1.5.0    [Build from master]

```
 python diagnose.py 
----------Python Info----------
('Version      :', '2.7.12')
('Compiler     :', 'GCC 5.4.0 20160609')
('Build        :', ('default', 'Nov 12 2018 14:36:49'))
('Arch         :', ('64bit', 'ELF'))
------------Pip Info-----------
('Version      :', '18.1')
('Directory    :', '/usr/local/lib/python2.7/dist-packages/pip')
----------MXNet Info-----------
('Version      :', '1.5.0')
('Directory    :', '/usr/local/lib/python2.7/dist-packages/mxnet-1.5.0-py2.7.egg/mxnet')
Hashtag not found. Not installed from pre-built package.
----------System Info----------
('Platform     :', 'Linux-4.4.0-141-generic-x86_64-with-Ubuntu-16.04-xenial')
('system       :', 'Linux')
('node         :', 'elara')
('release      :', '4.4.0-141-generic')
('version      :', '#167-Ubuntu SMP Wed Dec 5 10:40:15 UTC 2018')
----------Hardware Info----------
('machine      :', 'x86_64')
('processor    :', 'x86_64')
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                24
On-line CPU(s) list:   0-23
Thread(s) per core:    2
Core(s) per socket:    6
Socket(s):             2
NUMA node(s):          2
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 44
Model name:            Intel(R) Xeon(R) CPU           L5640  @ 2.27GHz
Stepping:              2
CPU MHz:               1596.000
CPU max MHz:           2261.0000
CPU min MHz:           1596.0000
BogoMIPS:              4533.49
Virtualisation:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              256K
L3 cache:              12288K
NUMA node0 CPU(s):     0-5,12-17
NUMA node1 CPU(s):     6-11,18-23
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 cx16 xtpr pdcm pcid dca sse4_1 sse4_2 popcnt aes lahf_lm epb kaiser tpr_shadow vnmi flexpriority ept vpid dtherm arat
----------Network Test----------
Setting timeout: 10
Timing for MXNet: https://github.com/apache/incubator-mxnet, DNS: 0.0195 sec, LOAD: 1.5077 sec.
Timing for PYPI: https://pypi.python.org/pypi/pip, DNS: 0.0316 sec, LOAD: 1.4635 sec.
Timing for FashionMNIST: https://apache-mxnet.s3-accelerate.dualstack.amazonaws.com/gluon/dataset/fashion-mnist/train-labels-idx1-ubyte.gz, DNS: 0.0617 sec, LOAD: 0.3168 sec.
Timing for Conda: https://repo.continuum.io/pkgs/free/, DNS: 0.0208 sec, LOAD: 1.1180 sec.
Timing for Gluon Tutorial(en): http://gluon.mxnet.io, DNS: 0.2185 sec, LOAD: 0.7843 sec.
Timing for Gluon Tutorial(cn): https://zh.gluon.ai, DNS: 0.1908 sec, LOAD: 0.7696 sec.
```

Package used (Python/R/Scala/Julia):
Python with GluonCV 0.4.0

MXNet commit hash:
17c8b25367859818b7aa5ec5466472f96f8c82e1

Build config:
docs/install/install_mxnet_ubuntu_python.sh
then a python setup.py install.

## Error Message:
No error message in logs etc, just when you ctrl-c the program (a kill is required to stop it).

## Minimum reproducible example
Just use gluoncv train-ssd.py:
gluon-cv/scripts/detection/ssd/train_ssd.py
try with num-workers 0 it'll work then num-workers > 0 it'll show the same symptoms I have shown here. 

## Steps to reproduce
working:
1. gluon-cv/scripts/detection/ssd/train_ssd.py --num-workers 0

not working:
1. gluon-cv/scripts/detection/ssd/train_ssd.py --num-workers 2

## What have you tried to solve it?

1. Tried pre-built binaries for linux system, I nuked cuda, cudann, python (all libraries too) and reinstalled all form scratch.

2. I even tried installing from source on MacBook with a 750M GPU and the same issue, on mac os, same versions built locally. Hence I am posting as it does not look like a local environment issue.

**Comment by lupesko** [comment]
@mxnet-label-bot please add [Bug]

**Comment by zachgk** [comment]
Thank you for submitting the issue! I'm labeling it so the MXNet community members can help resolve it.

@lupesko It seems like the label bot won't listen to you if you are too polite

@mxnet-label-bot add [Python, Gluon, Data-loading, Bug]

**Comment by zhreshold** [comment]
you might want to increase the shared memory size if it's small, or you can use thread_pool=True in DataLoader if you are using the latest mxnet on master

**Comment by Roshrini** [comment]
@evorta I ran gluon-cv/scripts/detection/ssd/train_ssd.py --num-workers 2 on Mac and also on ubuntu P2.8xlarge.. It seems to be running fine. Ran just 5 epochs though for testing. Can you check it again on MXNet master? Also did you get a chance to try fix suggested by @zhreshold?

**Comment by abhinavs95** [comment]
Looks like #14541 is a duplicate issue of this. Acc @zhreshold comment there the issue is fixed in 1.4.0 and master
